src/kernel/vmm.zig

```diff
@@ -0,0 +1,543 @@
+        ///     Bitmap.BitmapError.OutOfBounds - The physical or virtual addresses are out of bounds
+        ///     std.mem.Allocator.Error.OutOfMemory - Allocating the required memory failed
+        ///
+        pub fn set(self: *Self, virtual_start: usize, virtual_end: usize, physical_start: usize, physical_end: usize, attrs: Attributes) (VmmError || bitmap.Bitmap(u32).BitmapError || std.mem.Allocator.Error)!void {
+            var virt = virtual_start;
+            while (virt < virtual_end) : (virt += BLOCK_SIZE) {
+                if (try self.isSet(virt))
+                    return VmmError.AlreadyAllocated;
+            }
+            var phys = physical_start;
+            while (phys < physical_end) : (phys += BLOCK_SIZE) {
+                if (try pmm.isSet(phys))
+                    return VmmError.PhysicalAlreadyAllocated;
+            }
+            if (virtual_end - virtual_start != physical_end - physical_start)
+                return VmmError.PhysicalVirtualMismatch;
+            if (physical_start > physical_end)
+                return VmmError.InvalidPhysAddresses;
+            if (virtual_start > virtual_end)
+                return VmmError.InvalidVirtAddresses;
+
+            virt = virtual_start;
+            while (virt < virtual_end) : (virt += BLOCK_SIZE) {
+                try self.bmp.setEntry(virt / BLOCK_SIZE);
+            }
+
+            self.mapper.mapFn(virtual_start, virtual_end, physical_start, physical_end, attrs, self.allocator, self.payload);
+
+            var phys_list = std.ArrayList(usize).init(self.allocator);
+            phys = physical_start;
+            while (phys < physical_end) : (phys += BLOCK_SIZE) {
+                try pmm.setAddr(phys);
+                try phys_list.append(phys);
+            }
+            _ = try self.allocations.put(virt, Allocation{ .physical = phys_list });
+        }
+
+        ///
+        /// Allocate a number of contiguous blocks of virtual memory
+        ///
+        /// Arguments:
+        ///     INOUT self: *Self - The manager to allocate for
+        ///     IN num: u32 - The number of blocks to allocate
+        ///     IN attrs: Attributes - The attributes to apply to the mapped memory
+        ///
+        /// Return: ?usize
+        ///     The address at the start of the allocated region, or null if no region could be allocated due to a lack of contiguous blocks.
+        ///
+        /// Error: std.mem.Allocator.Error
+        ///     std.mem.AllocatorError.OutOfMemory: The required amount of memory couldn't be allocated
+        ///
+        pub fn alloc(self: *Self, num: u32, attrs: Attributes) std.mem.Allocator.Error!?usize {
+            // Ensure that there is both enough physical and virtual address space free
+            if (pmm.numFree() >= num and self.bmp.num_free_entries >= num) {
+                var block_list = std.ArrayList(usize).init(self.allocator);
```

Comment:
If moving the allocations after the bit map check, then won't need a arraylist

Reply:
We will need the list as we need to know which physical blocks correspond to the allocation so that they can be freed when the virtual allocation is freed.